### PR TITLE
Refactor Replace Repartition rule 

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes #.

# Rationale for this change
This PR refactors replace repartition rule implementation. With the new approach, the scope of the rule is extended. Rule no longer just replaces `RepartitionExec` with `SortPreservingRepartitionExec`. It replaces, `ExecThatLostOrdering` with `ExecVariantThatPreserveOrder`. 

Currently, we replace 
`RepartitionEXec` with `SortPreservingRepartitionEXec`
`CoalescePArtitionsExec` with `SortPreservingMergeExec`. 
In the future, if there are executors suitable for this conversion, this support can be extended easily. 

# What changes are included in this PR?

# Are these changes tested?
Yes, new tests are added, to test whether `CoalescePArtitionsExec` -> `SortPreservingMergeExec` conversion can be done successfully. 

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs [hunk 1]
```diff
@@ -15,136 +15,256 @@
...
+            .map(|item| item.plan.clone())
+            .collect();
+        let ordering_onwards = children_nodes
+            .into_iter()
+            .enumerate()
+            .map(|(idx, item)| {
+                // Leaves of the `coalesce_onwards` tree are `CoalescePartitionsExec`
+                // operators. This tree collects all the intermediate executors that
+                // maintain a single partition. If we just saw a `CoalescePartitionsExec`
+                // operator, we reset the tree and start accumulating.
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs:72]: This comment is outdated and confusing, let's remove and replace with an appropriate one.

**Review by ozankabak — APPROVED** [review]
LGTM except one minor outdated comment for which I left a note. @alamb or @wolffcm, can you please take a quick look too in case we missed something?

**Review by wolffcm — APPROVED** [review]
This seems quite reasonable to me.

File: datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs [hunk 2]
```diff
@@ -15,136 +15,256 @@
...
+    let ordering_onwards = &requirements.ordering_onwards;
+    if is_sort(plan) {
+        let exec_tree = if let Some(exec_tree) = &ordering_onwards[0] {
+            exec_tree
+        } else {
+            return Ok(Transformed::No(requirements));
+        };
+        // For unbounded cases, replace with the order-preserving variant in
+        // any case, as doing so helps fix the pipeline.
+        let is_unbounded = unbounded_output(plan);
```
[the remark this answers is not in the document]
Comment (ozankabak) [reply datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs:247]: AFAICT this call uses the local `unbounded_output` wrapper which suppresses such errors and replaces them with a hypothetical/pretend `true` result. I think that semantics is OK here.

File: datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs [hunk 3]
```diff
@@ -15,136 +15,256 @@
...
+        let updated_sort_input = get_updated_plan(
+            exec_tree,
+            is_spr_better | is_unbounded,
+            is_spm_better | is_unbounded,
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs:251]:
```suggestion
            is_spr_better || is_unbounded,
            is_spm_better || is_unbounded,
```

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 4]
```diff
@@ -367,11 +369,22 @@ impl PhysicalOptimizerRule for EnforceSorting {
...
+                )
+            })?;
+
         // Execute a top-down traversal to exploit sort push-down opportunities
         // missed by the bottom-up traversal:
-        let sort_pushdown = SortPushDown::init(new_plan);
+        let sort_pushdown = SortPushDown::init(updated_plan.plan);
         let adjusted = sort_pushdown.transform_down(&pushdown_sorts)?;
-        adjusted.plan.transform_down(&replace_repartition_execs)
+        // adjusted.plan.transform_down(&replace_repartition_execs)
```
[the remark this answers is not in the document]
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/sort_enforcement.rs:386]: Indeed, thanks

File: datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs:247]: Yes, code behaves as @ozankabak described. With this semantic, we postpone the validation of "whether a physical plan can run without braking pipeline" to `PipelineChecker`. During optimization, we do not restrict ourselves with such checks. What we care is that, whether relevant executor would receive unbounded data or not. If after all of the optimization, physical plan still cannot be run without braking pipeline. Then an error is raised indicating physical plan breaks pipeline.

[the remark this answers is not in the document]
Comment (wolffcm) [reply datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs:246]: Got it--I was confused. This makes sense.

**Review by alamb — COMMENTED** [review]
LGTM -- thanks @mustafasrepo , @ozankabak  and @wolffcm

**Pull request merged**